admin/services/generate-pins.service.js

```diff
@@ -37,28 +41,28 @@ const generatePinService = {
     pupils = await pupilDataService.find({ _id: { $in: ids } })
     // Apply the updates to the pupil object(s)
     pupils.forEach(pupil => {
-      if (!pupil.pin || !generatePinService.isValidPin(pupil.pinExpiresAt)) {
+      if (!generatePinService.isValidPin(pupil.pin, pupil.pinExpiresAt)) {
         const length = 5
         pupil.pin = generatePinService.generateRandomPin(length)
-        pupil.pinExpiresAt = moment.now()
+        pupil.pinExpiresAt = fourPMToday()
       }
     })
     return pupils
   },
 
   generateSchoolPassword: (school) => {
     let { schoolPin, pinExpiresAt } = school
-    if (!schoolPin || !generatePinService.isValidPin(pinExpiresAt)) {
+    if (!generatePinService.isValidPin(schoolPin, pinExpiresAt)) {
       const length = 8
       school.schoolPin = generatePinService.generateRandomPin(length)
-      school.pinExpiresAt = moment.now()
+      school.pinExpiresAt = fourPMToday()
     }
     return school
   },
 
-  isValidPin: (pinExpiresAt) => {
-    if (!pinExpiresAt) return false
-    return moment(pinExpiresAt).isBefore(moment().startOf('day').add(16, 'hours'))
+  isValidPin: (pin, pinExpiresAt) => {
+    if (!pinExpiresAt || !pin) return false
+    return moment(pinExpiresAt).isAfter(moment.utc())
```

Comment:
I would just pass the pupil into the method

Reply:
@GuyHarwood all I need really is these two properties though.